File: erlang/ch1/ex104.erl
%%%-------------------------------------------------------------------
%%% @author Yohan
%%% @copyright (C) 2018, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 27. Jul 2018 10:21 AM
%%%-------------------------------------------------------------------
-module(ex104).
-author("Yohan").

%% API
-export([main/0, a_plus_abs_b/2]).

main() ->
  io:fwrite("~w\n",[a_plus_abs_b(-1,-1)]),
  io:fwrite("~w\n",[a_plus_abs_b(-1,-1)]),
  io:fwrite("~w\n",[a_plus_abs_b(2,3)]),
  io:fwrite("~w\n",[a_plus_abs_b(2,-3)]),
  io:fwrite("~w\n",[a_plus_abs_b(-2,3)]),
  io:fwrite("~w\n",[a_plus_abs_b(-2,-3)]).


%%% Clearly unnecessary use of lambdas
a_plus_abs_b(A, B) when B < 0 -> (fun(X,Y)-> X-Y end)(A, B);
a_plus_abs_b(A, B) -> (fun(X,Y)-> X+Y end)(A, B).

File: julia/ch1/ex104.jl
#=
ex104:
- Julia version: 0.6.4
- Author: Yohan
- Date: 2018-07-31
=#

a_plus_abs_b_lambda(a, b) = ( b > 0 ? (x->a+x) : (x->a-x) )(b)

function a_plus_abs_b_operator(a, b)
   (if b > 0
       +
   else
       -
   end)(a,b)
end

a_plus_abs_b(a, b) = a_plus_abs_b_operator(a, b)
